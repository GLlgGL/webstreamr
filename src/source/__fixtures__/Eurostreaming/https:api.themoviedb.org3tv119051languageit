{"adult":false,"backdrop_path":"/qg8Gv2w0dDL8cMsG2QO2hWp58wy.jpg","created_by":[{"id":18923,"credit_id":"602e00776743fa003ce65015","name":"Miles Millar","original_name":"Miles Millar","gender":2,"profile_path":"/4phggQUKYg2WMMEy4JW8y6pKEVS.jpg"},{"id":18924,"credit_id":"602e0072d55697003f5a8765","name":"Alfred Gough","original_name":"Alfred Gough","gender":2,"profile_path":"/wZV1hProzjw9oR9HtjILDMn2K6A.jpg"}],"episode_run_time":[],"first_air_date":"2022-11-23","genres":[{"id":10765,"name":"Sci-Fi & Fantasy"},{"id":9648,"name":"Mistero"},{"id":35,"name":"Commedia"}],"homepage":"https://www.netflix.com/title/81231974","id":119051,"in_production":false,"languages":["en"],"last_air_date":"2025-09-03","last_episode_to_air":{"id":6165599,"name":"Guerra, triste guerra","overview":"Quando vengono alla luce segreti di famiglia a lungo sepolti, Mercoledì corre contro il tempo per impedire che un'oscura profezia si avveri.","vote_average":7.6,"vote_count":40,"air_date":"2025-09-03","episode_number":8,"episode_type":"finale","production_code":"","runtime":49,"season_number":2,"show_id":119051,"still_path":"/eqCAySjP9dXS4xgtbVAx1UjEi7D.jpg"},"name":"Mercoledì","next_episode_to_air":null,"networks":[{"id":213,"logo_path":"/wwemzKWzjKYJFfCeiB57q3r4Bcm.png","name":"Netflix","origin_country":""}],"number_of_episodes":16,"number_of_seasons":2,"origin_country":["US"],"original_language":"en","original_name":"Wednesday","overview":"Perspicace, sarcastica e... un po' inquietante, Mercoledì Addams indaga su misteri intricati mentre conosce nuovi amici e rivali alla Nevermore Academy.","popularity":195.3801,"poster_path":"/sci4aAuTCWQsIs2uo7LoZVtMdD0.jpg","production_companies":[{"id":2230,"logo_path":"/igOjospzsKQIbpuFCGhoN5F9icS.png","name":"MGM Television","origin_country":"US"},{"id":20356,"logo_path":"/awFYFsOyoG3VttUmWvnhgZgXxxh.png","name":"Millar Gough Ink","origin_country":"US"},{"id":186866,"logo_path":"/crmJkiAcLftaxF3GoQvRHPs7gyv.png","name":"Toluca Pictures","origin_country":"US"},{"id":8601,"logo_path":"/kqubCH3sfBAk4znUj7hjqv8ffgi.png","name":"Tim Burton Productions","origin_country":"US"}],"production_countries":[{"iso_3166_1":"US","name":"United States of America"}],"seasons":[{"air_date":"2022-11-23","episode_count":8,"id":182137,"name":"Stagione 1","overview":"","poster_path":"/v9fNsN3WNXObKJjBWkiKMuT3XoR.jpg","season_number":1,"vote_average":8.5},{"air_date":"2025-08-06","episode_count":8,"id":345672,"name":"Stagione 2","overview":"","poster_path":"/aamw6JjKwTy6bdviyIcFKekSD6x.jpg","season_number":2,"vote_average":7.6}],"spoken_languages":[{"english_name":"English","iso_639_1":"en","name":"English"}],"status":"Ended","tagline":"","type":"Scripted","vote_average":8.4,"vote_count":9838}