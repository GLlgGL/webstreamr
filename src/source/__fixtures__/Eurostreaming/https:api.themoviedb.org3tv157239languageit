{"adult":false,"backdrop_path":"/s7Lim1xma0qCNwlrQlaxM9mFLyN.jpg","created_by":[{"id":202816,"credit_id":"620bdfc73e01ea002b2f90b1","name":"Noah Hawley","original_name":"Noah Hawley","gender":2,"profile_path":"/pxVP1lywEBA9y8x5Xe3kM7m4ai3.jpg"}],"episode_run_time":[],"first_air_date":"2025-08-12","genres":[{"id":10765,"name":"Sci-Fi & Fantasy"},{"id":18,"name":"Dramma"}],"homepage":"https://www.fxnetworks.com/shows/alien-earth","id":157239,"in_production":true,"languages":["en"],"last_air_date":"2025-09-23","last_episode_to_air":{"id":6231390,"name":"I veri mostri","overview":"Emerge una nuova dinamica di potere.","vote_average":7.54,"vote_count":25,"air_date":"2025-09-23","episode_number":8,"episode_type":"finale","production_code":"XWYV1008","runtime":48,"season_number":1,"show_id":157239,"still_path":"/tssL6QBaKUkdXtmPUidrsREIEX5.jpg"},"name":"Alien - Pianeta Terra","next_episode_to_air":null,"networks":[{"id":88,"logo_path":"/aexGjtcs42DgRtZh7zOxayiry4J.png","name":"FX","origin_country":"US"},{"id":453,"logo_path":"/pqUTCleNUiTLAVlelGxUgWn1ELh.png","name":"Hulu","origin_country":"US"}],"number_of_episodes":8,"number_of_seasons":1,"origin_country":["US"],"original_language":"en","original_name":"Alien: Earth","overview":"Quando una misteriosa nave spaziale si schianta sulla Terra, una giovane donna (Sydney Chandler) e un gruppo di soldati tattici fanno una scoperta fatidica che li mette faccia a faccia con la più grande minaccia per il pianeta.","popularity":213.6674,"poster_path":"/y6DP95YbjwgDuveTNP28tFqAvWk.jpg","production_companies":[{"id":15990,"logo_path":"/5cT4zwHA66uNAr2p3CcBDLddXu2.png","name":"FX Productions","origin_country":"US"},{"id":221347,"logo_path":"/6Ry6uNBaa0IbbSs1XYIgX5DkA9r.png","name":"Scott Free Productions","origin_country":"US"},{"id":20814,"logo_path":null,"name":"26 Keys Productions","origin_country":"US"}],"production_countries":[{"iso_3166_1":"US","name":"United States of America"}],"seasons":[{"air_date":"2025-08-12","episode_count":8,"id":243129,"name":"Stagione 1","overview":"","poster_path":"/4rZdYuKbxakEGZRrgnPK7Vm5asb.jpg","season_number":1,"vote_average":7.2}],"spoken_languages":[{"english_name":"English","iso_639_1":"en","name":"English"}],"status":"Returning Series","tagline":"Eravamo più al sicuro nello spazio.","type":"Scripted","vote_average":7.616,"vote_count":565}