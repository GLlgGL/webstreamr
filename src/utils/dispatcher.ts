import { socksDispatcher } from 'fetch-socks';
import { Agent, Dispatcher, interceptors, ProxyAgent } from 'undici';

const createBasicDispatcher = (): Dispatcher => {
  if (process.env['ALL_PROXY']) {
    const proxyUrl = new URL(process.env['ALL_PROXY']);
    if (proxyUrl.protocol === 'socks5:') {
      return socksDispatcher({ type: 5, host: proxyUrl.hostname, port: parseInt(proxyUrl.port) }, { allowH2: true });
    } else {
      return new ProxyAgent({ uri: proxyUrl.href, allowH2: true });
    }
  } else {
    return new Agent({ allowH2: true });
  }
};

export const createDispatcher = (): Dispatcher => {
  return createBasicDispatcher().compose(
    interceptors.dns(),
    interceptors.retry({ maxRetries: 3 }),
  );
};
