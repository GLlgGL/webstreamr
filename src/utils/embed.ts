import { unpack } from 'unpacker';

export const unpackEval = (html: string): string => {
  const evalMatch = html.match(/eval\(function\(p,a,c,k,e,d\).*\)\)/);
  if (!evalMatch) {
    throw new Error(`No p.a.c.k.e.d string found`);
  }

  return unpack(evalMatch[0]);
};

export const extractUrlFromPacked = (html: string, linkRegExps: RegExp[]): URL => {
  const unpacked = unpackEval(html);

  for (const linkRegexp of linkRegExps) {
    const linkMatch = unpacked.match(linkRegexp);
    if (linkMatch && linkMatch[1]) {
      return new URL(`https://${linkMatch[1].replace(/^(https:)?\/\//, '')}`);
    }
  }

  throw new Error(`Could not find a stream link in embed`);
};
