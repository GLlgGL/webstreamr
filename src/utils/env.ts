import * as os from 'node:os';
import { Request } from 'express';

export const envGet = (name: string): string | undefined => process.env[name];

export const envGetRequired = (name: string): string => {
  const value = envGet(name);
  if (!value) {
    throw new Error(`Environment variable "${name}" is not configured.`);
  }

  return value;
};

export const envGetAppId = (): string => process.env['MANIFEST_ID'] || 'webstreamr';

export const envGetAppName = (): string => process.env['MANIFEST_NAME'] || 'WebStreamr';

export const envIsProd = (): boolean => process.env['NODE_ENV'] === 'production';

export const isElfHostedInstance = (req: Request): boolean => req.host.endsWith('elfhosted.com');

export const getCacheDir = (): string => envGet('CACHE_DIR') ?? os.tmpdir();
