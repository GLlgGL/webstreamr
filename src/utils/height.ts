import { Context } from '../types';
import { CustomRequestInit, Fetcher } from './Fetcher';

export const guessHeightFromPlaylist = async (ctx: Context, fetcher: Fetcher, url: URL, init?: CustomRequestInit): Promise<number | undefined> => {
  const m3u8Data = await fetcher.text(ctx, url, init);

  const heights = Array.from(m3u8Data.matchAll(/\d+x(\d+)|(\d+)p/g))
    .map(heightMatch => heightMatch[1] ?? heightMatch[2])
    .filter(height => height !== undefined)
    .map(height => parseInt(height));

  return heights.length ? Math.max(...heights) : /* istanbul ignore next */ undefined;
};
