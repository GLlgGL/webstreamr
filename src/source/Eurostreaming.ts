import * as cheerio from 'cheerio';
import { ContentType } from 'stremio-addon-sdk';
import { Context, CountryCode } from '../types';
import { Fetcher, getTmdbId, getTmdbNameAndYear, Id } from '../utils';
import { Source, SourceResult } from './Source';

export class Eurostreaming extends Source {
  public readonly id = 'eurostreaming';

  public readonly label = 'Eurostreaming';

  public readonly contentTypes: ContentType[] = ['series'];

  public readonly countryCodes: CountryCode[] = [CountryCode.it];

  public readonly baseUrl = 'https://eurostreaming.luxe';

  private readonly fetcher: Fetcher;

  public constructor(fetcher: Fetcher) {
    super();

    this.fetcher = fetcher;
  }

  public async handleInternal(ctx: Context, _type: string, id: Id): Promise<SourceResult[]> {
    const tmdbId = await getTmdbId(ctx, this.fetcher, id);
    const [name] = await getTmdbNameAndYear(ctx, this.fetcher, tmdbId, 'it');

    const seriesPageUrl = await this.fetchSeriesPageUrl(ctx, name);
    if (!seriesPageUrl) {
      return [];
    }

    const html = await this.fetcher.text(ctx, seriesPageUrl);

    const $ = cheerio.load(html);

    const title = `${name} ${tmdbId.season}x${tmdbId.episode}`;

    return Promise.all(
      $(`[data-num="${tmdbId.season}x${tmdbId.episode}"]`)
        .siblings('.mirrors')
        .children('[data-link!="#"]')
        .map((_i, el) => new URL($(el).attr('data-link') as string))
        .toArray()
        .filter(url => !url.host.match(/eurostreaming/))
        .map(url => ({ url, meta: { countryCodes: [CountryCode.it], referer: seriesPageUrl.href, title } })),
    );
  };

  private fetchSeriesPageUrl = async (ctx: Context, keyword: string): Promise<URL | undefined> => {
    const postUrl = new URL('/index.php?do=search', this.baseUrl);

    const form = new URLSearchParams();
    form.append('subaction', 'search');
    form.append('story', keyword);

    const html = await this.fetcher.textPost(
      ctx,
      postUrl,
      form.toString(),
      {
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
          'Referer': postUrl.origin,
        },
      },
    );

    const $ = cheerio.load(html);

    const exactKeyWordMatchUrl = $(`.post-thumb a[href][title="${keyword}"]:first`)
      .map((_i, el) => new URL($(el).attr('href') as string))
      .get(0);

    const partialKeyWordMatchUrl = $(`.post-thumb a[href][title*="${keyword}"]:first`)
      .map((_i, el) => new URL($(el).attr('href') as string))
      .get(0);

    return exactKeyWordMatchUrl ?? partialKeyWordMatchUrl;
  };
}
