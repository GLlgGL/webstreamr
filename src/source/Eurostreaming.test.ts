import { createTestContext } from '../test';
import { FetcherMock, TmdbId } from '../utils';
import { Eurostreaming } from './Eurostreaming';

const ctx = createTestContext({ it: 'on' });

describe('Eurostreaming', () => {
  let source: Eurostreaming;

  beforeEach(() => {
    source = new Eurostreaming(new FetcherMock(`${__dirname}/__fixtures__/Eurostreaming`));
  });

  test('handles non-existent series gracefully', async () => {
    const streams = await source.handle(ctx, 'series', new TmdbId(61945, 1, 1));
    expect(streams).toHaveLength(0);
  });

  test('handle imdb black mirror s2e4', async () => {
    const streams = await source.handle(ctx, 'series', new TmdbId(42009, 2, 4));
    expect(streams).toMatchSnapshot();
  });

  test('handle lost s1e1', async () => {
    const streams = await source.handle(ctx, 'series', new TmdbId(4607, 1, 1));
    expect(streams).toMatchSnapshot();
  });

  test('last of us s1e1', async () => {
    const streams = await source.handle(ctx, 'series', new TmdbId(100088, 1, 1));
    expect(streams).toMatchSnapshot();
  });

  test('game of thrones s1e1', async () => {
    const streams = await source.handle(ctx, 'series', new TmdbId(1399, 1, 1));
    expect(streams).toMatchSnapshot();
  });

  test('wednesday s1e1', async () => {
    const streams = await source.handle(ctx, 'series', new TmdbId(119051, 1, 1));
    expect(streams).toMatchSnapshot();
  });
});
