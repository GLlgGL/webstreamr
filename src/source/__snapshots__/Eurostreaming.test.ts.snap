// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Eurostreaming game of thrones s1e1 1`] = `
[
  {
    "meta": {
      "countryCodes": [
        "it",
      ],
      "referer": "https://eurostreaming.luxe/stream/194-visualizza-game-of-thrones-il-trono-di-spade-5-streaming.html",
      "title": "Il Trono di Spade 1x1",
    },
    "url": "https://supervideo.cc/e/05jfivnr5hiw",
  },
  {
    "meta": {
      "countryCodes": [
        "it",
      ],
      "referer": "https://eurostreaming.luxe/stream/194-visualizza-game-of-thrones-il-trono-di-spade-5-streaming.html",
      "title": "Il Trono di Spade 1x1",
    },
    "url": "https://dropload.io/embed-dub8oh3k3yk5.html",
  },
]
`;

exports[`Eurostreaming handle imdb black mirror s2e4 1`] = `
[
  {
    "meta": {
      "countryCodes": [
        "it",
      ],
      "referer": "https://eurostreaming.luxe/stream/284-black-mirror-streaming-streaming.html",
      "title": "Black Mirror 2x4",
    },
    "url": "https://supervideo.cc/e/anf0clp2vb6y",
  },
  {
    "meta": {
      "countryCodes": [
        "it",
      ],
      "referer": "https://eurostreaming.luxe/stream/284-black-mirror-streaming-streaming.html",
      "title": "Black Mirror 2x4",
    },
    "url": "https://dropload.io/embed-978xyyxi3ldq.html",
  },
]
`;

exports[`Eurostreaming handle lost s1e1 1`] = `
[
  {
    "meta": {
      "countryCodes": [
        "it",
      ],
      "referer": "https://eurostreaming.luxe/stream/496-lost-streaming.html",
      "title": "Lost 1x1",
    },
    "url": "https://supervideo.cc/e/lgslt2lst4jv",
  },
  {
    "meta": {
      "countryCodes": [
        "it",
      ],
      "referer": "https://eurostreaming.luxe/stream/496-lost-streaming.html",
      "title": "Lost 1x1",
    },
    "url": "https://dropload.io/embed-vabynwo8glxz.html",
  },
]
`;

exports[`Eurostreaming last of us s1e1 1`] = `
[
  {
    "meta": {
      "countryCodes": [
        "it",
      ],
      "referer": "https://eurostreaming.luxe/stream/2605-the-last-of-us-streamingcommunity-streaming.html",
      "title": "The Last of Us 1x1",
    },
    "url": "https://supervideo.cc/e/ic1mmvrh5xv6",
  },
  {
    "meta": {
      "countryCodes": [
        "it",
      ],
      "referer": "https://eurostreaming.luxe/stream/2605-the-last-of-us-streamingcommunity-streaming.html",
      "title": "The Last of Us 1x1",
    },
    "url": "https://dropload.io/embed-uyzfrh62n31l.html",
  },
]
`;

exports[`Eurostreaming wednesday s1e1 1`] = `
[
  {
    "meta": {
      "countryCodes": [
        "it",
      ],
      "referer": "https://eurostreaming.luxe/stream/2529-mercoledi-wednesday-streaming.html",
      "title": "Mercoledì 1x1",
    },
    "url": "https://supervideo.cc/e/n6i7qyxpkw2e",
  },
  {
    "meta": {
      "countryCodes": [
        "it",
      ],
      "referer": "https://eurostreaming.luxe/stream/2529-mercoledi-wednesday-streaming.html",
      "title": "Mercoledì 1x1",
    },
    "url": "https://dropload.tv/e/fkbw1uszc0nh",
  },
]
`;
